File: Sources/structy/06_graph/ShortestPath.swift
//
//  ShortestPath.swift
//  MyCLI
//
//  Created by new on 3/31/26.
//

func shortestPath(_ edges: [[String]], _ nodeA: String, _ nodeB: String) -> Int {
    var G: Dictionary<String, Set<String>> = [:]
    for edge in edges {
        let from = edge[0]
        let to = edge[1]
        G[from, default: Set<String>()].insert(to)
        G[to, default: Set<String>()].insert(from)
    }

    let q = Queue<(String, Int)>()
    var visited = Set<String>()
    q.enqueue((nodeA, 0))
    visited.insert(nodeA)
    while !q.isEmpty {
        guard let (v, d) = q.dequeue() else {
            continue
        }

        if v == nodeB {
            return d
        }


        for next_v in G[v, default: Set<String>()] {
            if visited.contains(next_v) {
                continue
            }

            q.enqueue((next_v, d + 1))
        }
    }

    return -1
}

File: Sources/structyC/06_graph/shortest_path.cpp
//
//  shortest_path.cpp
//  MyCLI
//
//  Created by new on 3/31/26.
//

#include <vector>
#include <tuple>
#include <string>
#include <queue>
#include <unordered_map>
#include <unordered_set>

int shortestPath(std::vector<std::tuple<std::string, std::string>> edges,
                 std::string nodeA,
                 std::string nodeB) {
    std::unordered_map<std::string, std::unordered_set<std::string>> G{};

    for (std::tuple<std::string, std::string> e : edges) {
        // or
        // auto [from, to] = e;
        auto from = std::get<0>(e);
        auto to = std::get<1>(e);
        if (G.find(from) == G.end()) {
            G[from] = std::unordered_set<std::string>{};
        }
        if (G.find(to) == G.end()) {
            G[to] = std::unordered_set<std::string>{};
        }
        G[from].insert(to);
        G[to].insert(from);
    }

    std::queue<std::string> q{};
    q.push(nodeA);
    std::unordered_set<std::string> visited{};
    visited.insert(nodeA);
    int length = 0;

    while (!q.empty()) {
        auto level = q.size();
        for (auto i=0;i<level;i++) {
            std::string v = q.front();
            q.pop();
            if (v == nodeB) {
                return length;
            }
            for (std::string next_v : G[v]) {
                if (visited.find(next_v) != visited.end()) {
                    continue;
                }
                visited.insert(next_v);
                q.push(next_v);
            }
        }
        length++;
    }
    return -1;
}


